services/gundeck/src/Gundeck/Push/Native.hs

```diff
@@ -30,13 +31,30 @@ import qualified Gundeck.Notification.Data as Stream
 import qualified Gundeck.Push.Data         as Data
 import qualified System.Logger.Class       as Log
 
-push :: NativePush -> [Address] -> Gundeck [Result]
+push :: NativePush -> [Address] -> Gundeck [()]
 push _    [] = return []
 push m   [a] = pure <$> push1 m a
 push m addrs = mapConcurrently (push1 m) addrs
```

Comment:
I don't think you're actually solving the root cause here. We're still starting all the requests concurrently.  I think the budgeting should happen _outside_ of the `mapConcurrently`. not _inside_

Reply:
no, we want to limit the number of concurrent calls to push1.

however, you are correct in that this doesn't solve the issue:

1. Budget works across instances, but if `withBudget` is called twice concurrently, both callers with get the same budget unit to spend.  This may not be a problem when doing sms budgets per user, but it will be if we have many 1k (10k?  more?) calls per second.
2. It doesn't distinguish between the long-running calls that time out because SNS is not responding (not even with a "connection refused") and the calls that return successfully after a few milliseconds.  The former need to be limited much more aggressively than the latter.